src/os/bluestore/BlueStore.cc

```diff
@@ -639,9 +639,9 @@ struct Int64ArrayMergeOperator : public KeyValueDB::MergeOperator {
     ceph_assert(llen == rlen);
     ceph_assert((rlen % 8) == 0);
     new_value->resize(rlen);
-    const __le64* lv = (const __le64*)ldata;
-    const __le64* rv = (const __le64*)rdata;
-    __le64* nv = &(__le64&)new_value->at(0);
+    const ceph_le64* lv = (const ceph_le64*)ldata;
+    const ceph_le64* rv = (const ceph_le64*)rdata;
+    ceph_le64* nv = &(ceph_le64&)new_value->at(0);
```

Comment:
Do I understand correctly that this is a breaking change for any big endian bluestore instance('s stats)?  Given the other endian issues that this series fixes, is such a thing possible?

Reply:
I believe the current code is just self-inconsistent.  For these stats, it seems they're e.g. written via the encode method of volatile_statfs, which I believe is using the old-style encoding.h encoders, which actually do byte-swap.  This means that the merge routine (which didn't byte-swap before this fix) would simply access the values incorrectly.

But in general, of course this set of changes does lead to incompatible changes in objects written by some of the patches routines (in particular, everything using the new-style denc.h encoders).

Given that the code was self-inconsistent, which actually led to crashes we observed when simply running on big-endian systems alone, I believe at this point we should simply fix everything to byte-swap so that all objects are encoded as little-endian as intended -- I don't think it would be worthwhile (or even possible) to try to accommodate some theoretical existing installation on a big-endian system that was broken enough to write (some) big-endian encoded objects, but for whatever reason not broken enough to crash due to self-inconsistency.

Reply:
That's what I was hoping to hear :).  Sounds good!